TileDB-Inc/TileDB-Cloud-Py
# Notebook uploader

Ready for review, with the following caveat:

* We might decide how much of the auto-increment logic to port (or not to port) from `TileDB-Cloud-Jupyter-Contents`.
*  If destination exists, three options:
    * Reject the write with 'already exists'. 
    * Auto-increment (as done in Jupyter)
    * Overwrite current version (should not be the default; perhaps should be supported as an option)
* Today's code does the first option and can be merged if everyone's OK with that for now.

# Manual test scenario

* Use `download-notebook.py` (below) to get valid JSON
* Use `upload-notebooy.py`
* Click through in the TileDB Cloud console and verify all tabs work for the newly created notebook

```
$ cat download-notebook.py
#!/usr/bin/env python

# ================================================================
# Setup:
# * https://console.dev.tiledb.io
# * Set my storage path and storage credential name in Settings
#
# Note: cannot be invoked from within the base directory of either the
# TileDB-Py or TileDB-Cloud-Py repos -- or any directory with a subdirectory
# named `tiledb` -- since `import tiledb` / `import tiledb.cloud` clashes
# with the existence of that subdirectory.
# ================================================================

import sys
import tiledb.cloud

config = tiledb.Config()
config["rest.username"] = "johnkerl"
config["rest.password"] = "redacted"
# API endpoint used by https://console.dev.tiledb.io:
config["rest.server_address"] = "api.dev.tiledb.io"

# This is the array-URI format in TileDB Cloud:
tiledb_uri              = "tiledb://seth/Voila-Covid-Demo"
storage_credential_name = "my-sandbox-creds"
ipynb_file_name         = "/Users/johnkerl/Desktop/PLAN/foo.ipynb"

if len(sys.argv) == 3:
    tiledb_uri      = sys.argv[1]
    ipynb_file_name = sys.argv[2]

tiledb.cloud.download_ipnyb_file_name_from_cloud(
    tiledb_uri,
    storage_credential_name,
    ipynb_file_name,
)

print("Wrote "+ipynb_file_name)
```

```
$ cat upload-notebook.py
#!/usr/bin/env python

# ================================================================
# Setup:
# * https://console.dev.tiledb.io
# * Set my storage path and storage credential name in Settings
#
# Note: cannot be invoked from within the base directory of either the
# TileDB-Py or TileDB-Cloud-Py repos -- or any directory with a subdirectory
# named `tiledb` -- since `import tiledb` / `import tiledb.cloud` clashes
# with the existence of that subdirectory.
# ================================================================

import sys
import tiledb.cloud

tiledb.cloud.login(
    host="https://api.dev.tiledb.io/v1",
    username="johnkerl",
    password="redacted",
)

ipynb_file_name         = "/Users/johnkerl/Desktop/PLAN/foo.ipynb"
namespace               = "johnkerl"
array_name              = "testing-upload-014"
storage_path            = "s3://tiledb-johnkerl"
storage_credential_name = "my-sandbox-creds"

if len(sys.argv) == 3:
    ipynb_file_name = sys.argv[1]
    array_name      = sys.argv[2]

tiledb_uri = tiledb.cloud.upload_ipnyb_file_name_to_cloud(
    ipynb_file_name,
    namespace,
    array_name,
    storage_path,
    storage_credential_name,
)

print("tiledb_uri="+tiledb_uri)
```

Be sure to run this for the cases when the destination array does not already exist, and does already exist.

# Tracking

CH 7290
